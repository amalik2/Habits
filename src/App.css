@font-face {
  font-family: 'Jura';
  src: url(content/Jura-Regular.ttf) format("truetype");
}

.App {
  text-align: center;
  font-family: Jura;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
  font-family: Jura;
}

.App-title {
  font-size: 1.5em;
  font-family: Jura;
}

.App-intro {
  font-size: large;
  font-family: Jura;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}
