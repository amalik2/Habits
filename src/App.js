import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

import {isAlphanumeric} from './utilities/stringutilities';
import {getQueryable} from './utilities/queryutilities';

import HomePage from './views/home';

import User from './models/user';

/**
 * Main application view, allowing the user to sign in to the system
 */
class App extends Component {
	
	constructor(props){
		super(props);
		this.state = {
			// {User} logged in user
			loggedIn: null,
			// {String} the current entry in the name input field
			name: ""
		};
		this.handleSubmit = this.handleSubmit.bind(this);
		this.nameChange = this.nameChange.bind(this);
		this.signOut = this.signOut.bind(this);
		
		getQueryable();
	}
	
	/**
	 * Handles the user selecting the sign in button
	 * @param event - the button click event that led to this being called
	 */
	handleSubmit(event){
		
		if (this.state.name.length === 0){
			alert("Empty name");
			return;
		}
		
		if (!isAlphanumeric(this.state.name)){
			alert("Invalid name selected. Names must be alphanumeric.");
			return;
		}
		
		this.setState({
			loggedIn: new User(this.state.name)
		});
	}
	
	/**
	 * Handles the user changing the name input field text
	 * @param event - the input change event that triggered this function call
	 */
	nameChange(event){
		this.setState({
			name: event.target.value
		});
	}
	
	/**
	 * Signs the user out of the system
	 */
	signOut(){
		this.setState({
			loggedIn: null
		});
	}
	
	/**
	 * @return the header that will be displayed across all pages in the application
	 */
	getHeader(){
		return (
			<header className="App-header">
			  <img src={logo} className="App-logo" alt="logo" />
			  <h1 className="App-title">Habit Tracker</h1>
			</header>
		);
	}
	
	render() {
		if (this.state.loggedIn != null){
		  return (
				<div align='center'>
					{this.getHeader()}
					<HomePage user={this.state.loggedIn} signOut={this.signOut} />
				</div>
			);
		} else {
		  return (
			  <div className="App">
				{this.getHeader()}
				<form onSubmit={this.handleSubmit}>
					<label>
						<input type="text" name="Username" value={this.state.name} onChange={this.nameChange}/>
					</label>
					<input type="submit" value="Sign In" />
				</form>
			  </div>
			);
		}
	}
}

export default App;
