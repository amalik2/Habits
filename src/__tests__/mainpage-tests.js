import React from 'react';
import App from '../App';
import renderer from 'react-test-renderer';

import { mount } from 'enzyme';

/**
 * UI tests for the main page of the application
 */

test("Can't sign in with an invalid name", () => {
	const component = mount(
		<App />,
	);
  
	expect(component.state().loggedIn).toBe(null);
	expect(component.state().name).toBe("");
  
	const usernameForm = component.find('input');
	const usernameInput = usernameForm.first();
	const submit = usernameForm.at(1);
	
	const INVALID_SYMBOLS = "'.\"[]/?=-_+\|}{}`\\ ";
	
	// prevent logging in with an empty name
	submit.simulate('submit');
	expect(component.state().loggedIn).toBe(null);
	
	// make sure all invalid symbols are rejected
	for (let i = 0; i < INVALID_SYMBOLS.length; ++i){
		usernameInput.simulate('change', {target: {value: INVALID_SYMBOLS[i]}});
		expect(component.state().name).toBe(INVALID_SYMBOLS[i]);
		submit.simulate('submit');
		expect(component.state().loggedIn).toBe(null);
	}
	
	// make sure all of the following names work
	const VALID_NAMES = ["abc", "123", "abc123", " ab", " ab c ", "xyz 12345678910111213", "abcdefghijklmnopqrstuvwxyz1234567890"];
	const length = VALID_NAMES.length;
	for (let i = 0; i < length; ++i){
		usernameInput.simulate('change', {target: {value: VALID_NAMES[i]}});
		expect(component.state().name).toBe(VALID_NAMES[i]);
		submit.simulate('submit');
		expect(component.state().loggedIn).toBe(VALID_NAMES[i]);
	}
	
	
});

test("Error signing in or out of the application", () => {
	const component = mount(
		<App />,
	);
	
	expect(component.find('HomePage').length).toBe(0);
	
	const usernameForm = component.find('input');
	const usernameInput = usernameForm.first();
	const submit = usernameForm.at(1);
	
	usernameInput.simulate('change', {target: {value: "signintest"}});
	expect(component.state().name).toBe("signintest");
	submit.simulate('submit');
	expect(component.state().loggedIn).toBe("signintest");
	
	// make sure sign in takes user to the home page
	const home = component.find('HomePage').first();
	expect(home).not.toBe(null);
	
	// test to make sure the logout button works
	const buttons = home.find("Sidebar").find("button");
	// positon of logout is the last one in the list
	buttons.at(buttons.length - 1).simulate('click');
	
});
