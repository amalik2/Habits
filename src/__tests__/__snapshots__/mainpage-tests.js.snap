// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can't sign in with an invalid name 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Habit Tracker
    </h1>
  </header>
  <form
    onSubmit={[Function]}
  >
    <label>
      <input
        name="Username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Sign In"
    />
  </form>
</div>
`;

exports[`Link changes the class when hovered 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Habit Tracker
    </h1>
  </header>
  <form
    onSubmit={[Function]}
  >
    <label>
      <input
        name="Username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Sign In"
    />
  </form>
</div>
`;
